File: etude12/Java/Ant.java
import System;
import System.Collections.Generic;
public class Tuple<X, Y> { 
    public final X x; 
    public final Y y; 
    public Tuple(X x, Y y) { 
      this.x = x; 
      this.y = y; 
    } 
  } 
    public class Ant
    {
        private Dictionary<char, (string dir, string nState)> dna = new Dictionary<char, (string dir, string nState)>();
        private Dictionary< Tuple<X, Y>, char> plane = new Dictionary<Tuple<X, Y>, char>(); //key: position, vlaue: if it has been visited the state

        private int lastPos = 0; 
        private char defaultSym;


        //Constructor
        public Ant(Dictionary<char, (string dir, string nState)> dna, char defaultSym)
        {
            this.dna = dna;
            this.defaultSym = defaultSym;
        }

        public Tuple<X,Y> move(Tuple<X, Y> currentPos){
            char nextpos = 'S';
            if(plane.ContainsKey(currentPos)){
                //has already been walked over
                (string dir, string nState) dnaResult = dna[plane[currentPos]]; //dont need to check because plane[currentpos] will only return valid symbol coz it was set

                
                nextpos = dnaResult.dir[lastPos]; //where to go next
                plane[currentPos] = dnaResult.nState[lastPos]; //changes plane state accordingly
            }else{
                //state hasn't been visited before
                (string dir, string nState) dnaResult = dna[defaultSym];
                nextpos = dnaResult.dir[lastPos]; //where to go next
                plane.Add(currentPos,dnaResult.nState[lastPos]); //changes plane state accordingly
            }
            

            return getNewPos(nextpos, currentPos);
        }

        private (int, int) getNewPos(char dir, (int x, int y) currentPos){
            (int x,int y) newPos = (0,0);
            switch (dir)
            {
                case 'N':
                    //moving north
                    newPos.x = currentPos.x;
                    newPos.y = currentPos.y+1;
                    lastPos = 0;
                    break;
                case 'E':
                    //moving east
                    newPos.x = currentPos.x+1;
                    newPos.y = currentPos.y;
                    lastPos = 1;
                    break;
                case 'S':
                    //moving south
                    newPos.x = currentPos.x;
                    newPos.y = currentPos.y-1;
                    lastPos = 2;
                    break;
                case 'W':
                    //moving north
                    newPos.x = currentPos.x-1;
                    newPos.y = currentPos.y;
                    lastPos = 3;
                    break;
                default:
                    Console.WriteLine("Direction not valid");
                    break;
            }
            return newPos;
        }
    }
}

File: etude12/Ant.py


class Ant:

    dna = {}
    plane = {} # key: position, vlaue: if it has been visited the state

    lastPos = 0
    defaultSym = ""
    # Constructor
    def __init__(self, dna, defaultSym):
        self.dna = dna
        self.defaultSym = defaultSym
        self.plane = {}






    def getNewPos(self, dir, currentPos):
        newPos = (0, 0)
        x = 0
        y = 0
        if dir == 'N':
        # moving north
            x = currentPos[0]
            y = currentPos[1] + 1
            self.lastPos = 0

        elif dir == 'E':
            #moving east
            x = currentPos[0] +1
            y = currentPos[1]
            self.lastPos = 1

        elif dir == 'S':
            x = currentPos[0]
            y = currentPos[1] - 1
            self.lastPos = 2

        elif dir == 'W':
            x = currentPos[0] - 1
            y = currentPos[1]
            self.lastPos = 3
        else:
            print("Direction '{}' not valid".format(dir))

        return (x, y)

    def move(self, currentPos):
        nextpos = 'S'
        if currentPos in self.plane:
            #   has already been walked over
            dnaResult = self.dna[self.plane[currentPos]] # dont need to check because plane[currentpos] will onlyreturn valid symbol coz it was set

            nextpos = dnaResult[0][self.lastPos] #where to go next
            self.plane[currentPos] = dnaResult[1][self.lastPos] # changes plane state accordingly

        else :
            #state hasn't been visited before
            dnaResult = self.dna[self.defaultSym]
            nextpos = dnaResult[0][self.lastPos]# where to go next
            self.plane[currentPos] = dnaResult[1][self.lastPos] # changes plane state accordingly

        return self.getNewPos(nextpos, currentPos)

